arwen/hosttest/execution_gas_test.go

```diff
@@ -389,6 +391,148 @@ func TestGasUsed_AsyncCall(t *testing.T) {
+					test.CreateTransferEntry(test.ParentAddress, test.ChildAddress).
+						WithData(asyncCallData.ToBytes()).
+						WithGasLimit(gasForAsyncCall).
+						WithGasLocked(testConfig.GasLockCost).
+						WithCallType(vmcommon.AsynchronousCall).
+						WithValue(big.NewInt(testConfig.TransferFromParentToChild)),
+				)
+		})
+
+	childAsyncReturnData := [][]byte{{0}, []byte("thirdparty"), []byte("vault")}
+
+	/*
+		async cross shard parent -> child
+	*/
+	test.BuildMockInstanceCallTest(t).
+		WithContracts(
+			test.CreateMockContractOnShard(test.ChildAddress, 1).
+				WithBalance(testConfig.ChildBalance).
+				WithConfig(testConfig).
+				WithMethods(contracts.TransferToThirdPartyAsyncChildMock),
+		).
+		WithInput(test.CreateTestContractCallInputBuilder().
+			WithCallerAddr(test.ParentAddress).
+			WithRecipientAddr(test.ChildAddress).
+			WithGasProvided(gasForAsyncCall).
+			WithFunction(contracts.AsyncChildFunction).
+			WithArguments(
+				big.NewInt(testConfig.TransferToThirdParty).Bytes(),
+				[]byte(contracts.AsyncChildData),
+				asyncChildArg).
+			WithCallType(vmcommon.AsynchronousCall).
+			Build()).
+		WithSetup(func(host arwen.VMHost, world *worldmock.MockWorld) {
+			world.SelfShardID = 1
+			world.CurrentBlockInfo.BlockRound = 1
+			setZeroCodeCosts(host)
+			setAsyncCosts(host, testConfig.GasLockCost)
+		}).
+		AndAssertResults(func(world *worldmock.MockWorld, verify *test.VMOutputVerifier) {
+			verify.
+				Ok().
+				GasUsed(test.ChildAddress, gasUsedByChild).
+				GasRemaining(0).
+				ReturnData(childAsyncReturnData...).
+				Transfers(
+					test.CreateTransferEntry(test.ChildAddress, test.ThirdPartyAddress).
+						WithData([]byte(contracts.AsyncChildData)).
+						WithValue(big.NewInt(testConfig.TransferToThirdParty)),
+					test.CreateTransferEntry(test.ChildAddress, test.VaultAddress).
+						WithData([]byte{}).
+						WithValue(big.NewInt(testConfig.TransferToVault)),
+					test.CreateTransferEntry(test.ChildAddress, test.ParentAddress).
+						WithData(computeReturnDataForCallback(vmcommon.Ok, childAsyncReturnData)).
+						WithGasLimit(gasForAsyncCall-gasUsedByChild).
+						WithCallType(vmcommon.AsynchronousCallBack).
+						WithValue(big.NewInt(0)),
+				)
+		})
+
+	/*
+		async cross shard callback child -> parent
+	*/
+	test.BuildMockInstanceCallTest(t).
+		WithContracts(parentContract).
+		WithInput(test.CreateTestContractCallInputBuilder().
+			WithCallerAddr(test.ChildAddress).
+			WithRecipientAddr(test.ParentAddress).
+			WithGasProvided(gasForAsyncCall-gasUsedByChild+asyncBaseTestConfig.GasLockCost).
+			WithFunction("callBack").
+			WithArguments([]byte{}, []byte{0}, []byte("thirdparty"), []byte("vault")).
+			WithCallType(vmcommon.AsynchronousCallBack).
+			Build()).
+		WithSetup(func(host arwen.VMHost, world *worldmock.MockWorld) {
+			world.SelfShardID = 0
+			world.CurrentBlockInfo.BlockRound = 2
+			accountHandler, _ := world.GetUserAccount(test.ParentAddress)
```

Comment:
The following comment would help:
```suggestion
                        // Mock the storage as if the parent was already executed
			accountHandler, _ := world.GetUserAccount(test.ParentAddress)
```